client/trino-client/src/main/java/io/trino/client/spooling/encoding/JsonDecodingUtils.java

```diff
@@ -0,0 +1,494 @@
+import static io.trino.client.ClientStandardTypes.TIME_WITH_TIME_ZONE;
+import static io.trino.client.ClientStandardTypes.TINYINT;
+import static io.trino.client.ClientStandardTypes.UUID;
+import static io.trino.client.ClientStandardTypes.VARCHAR;
+import static java.lang.String.format;
+import static java.util.Collections.unmodifiableList;
+import static java.util.Collections.unmodifiableMap;
+import static java.util.Objects.requireNonNull;
+
+public final class JsonDecodingUtils
+{
+    private JsonDecodingUtils() {}
+
+    static TypeDecoder[] createTypeDecoders(List<Column> columns)
+    {
+        verify(!columns.isEmpty(), "Columns must not be empty");
+        return columns.stream()
+                .map(column -> createTypeDecoder(column.getTypeSignature()))
+                .toArray(TypeDecoder[]::new);
+    }
+
+    interface TypeDecoder
+    {
+        Object decode(JsonParser parser)
+                throws IOException;
+    }
+
+    private static TypeDecoder createTypeDecoder(ClientTypeSignature signature)
+    {
+        switch (signature.getRawType()) {
+            case BIGINT:
+                return new BigIntegerDecoder();
+            case INTEGER:
+                return new IntegerDecoder();
+            case SMALLINT:
+                return new SmallintDecoder();
+            case TINYINT:
+                return new TinyintDecoder();
+            case DOUBLE:
+                return new DoubleDecoder();
+            case REAL:
+                return new RealDecoder();
+            case BOOLEAN:
+                return new BooleanDecoder();
+            case ARRAY:
+                return new ArrayDecoder(signature);
+            case MAP:
+                return new MapDecoder(signature);
+            case ROW:
+                return new RowDecoder(signature);
+            case VARCHAR:
+            case JSON:
+            case TIME:
+            case TIME_WITH_TIME_ZONE:
+            case TIMESTAMP:
+            case TIMESTAMP_WITH_TIME_ZONE:
+            case DATE:
+            case INTERVAL_YEAR_TO_MONTH:
+            case INTERVAL_DAY_TO_SECOND:
+            case IPADDRESS:
+            case UUID:
+            case DECIMAL:
+            case CHAR:
+            case GEOMETRY:
+            case SPHERICAL_GEOGRAPHY:
+                return new StringDecoder();
+            default:
+                return new Base64Decoder();
+        }
+    }
+
+    private static class BigIntegerDecoder
+            implements TypeDecoder
+    {
+        @Override
+        public Object decode(JsonParser parser)
+                throws IOException
+        {
+            switch (parser.currentToken()) {
+                case VALUE_NULL:
+                    return null;
+                case VALUE_NUMBER_INT:
+                    return parser.getLongValue();
+                case VALUE_STRING:
+                    return Long.parseLong(parser.getValueAsString());
```

Comment:
was it like that before? Why do we have two on-wire representations?

Reply:
Yes, you can check `FixJsonUtils`:

```
case BIGINT:
                return (value) -> {
                    if (value instanceof String) {
                        return Long.parseLong((String) value);
                    }
                    return ((Number) value).longValue();
                };
```